torch/utils/data/sampler.py

```diff
@@ -114,6 +114,7 @@ def __iter__(self):
         if self.generator is None:
             generator = torch.Generator()
             generator.manual_seed(int(torch.empty((), dtype=torch.int64).random_().item()))
+            self.generator = generator
         else:
             generator = self.generator
```

Comment:
This part can be eliminated. And the following generator can be also replaced by self.generator.

Reply:
You are right.

In addition, would it be more appropriate if we moved the following `if` code into `__init__` of `RandomSampler`?  That is, `self.generator` is instantiated only when the class `RandomSampler` is initialized.

i.e., move these 
 https://github.com/pytorch/pytorch/blob/5f2fb58531c1ed472e979f82aa870f7cf50ce852/torch/utils/data/sampler.py#L114-L117

to

https://github.com/pytorch/pytorch/blob/5f2fb58531c1ed472e979f82aa870f7cf50ce852/torch/utils/data/sampler.py#L86-L91

Reply:
Look like this:

```python
class RandomSampler(Sampler[int]):
    r"""Samples elements randomly. If without replacement, then sample from a shuffled dataset.
    If with replacement, then user can specify :attr:`num_samples` to draw.
    Args:
        data_source (Dataset): dataset to sample from
        replacement (bool): samples are drawn on-demand with replacement if ``True``, default=``False``
        num_samples (int): number of samples to draw, default=`len(dataset)`. This argument
            is supposed to be specified only when `replacement` is ``True``.
        generator (Generator): Generator used in sampling.
    """
    data_source: Sized
    replacement: bool

    def __init__(self, data_source: Sized, replacement: bool = False,
                 num_samples: Optional[int] = None, generator=None) -> None:
        self.data_source = data_source
        self.replacement = replacement
        self._num_samples = num_samples
        if generator is None:
            generator = torch.Generator()
            generator.manual_seed(int(torch.empty((), dtype=torch.int64).random_().item()))
        self.generator = generator

        if not isinstance(self.replacement, bool):
            raise TypeError("replacement should be a boolean value, but got "
                            "replacement={}".format(self.replacement))

        if self._num_samples is not None and not replacement:
            raise ValueError("With replacement=False, num_samples should not be specified, "
                             "since a random permute will be performed.")

        if not isinstance(self.num_samples, int) or self.num_samples <= 0:
            raise ValueError("num_samples should be a positive integer "
                             "value, but got num_samples={}".format(self.num_samples))

    @property
    def num_samples(self) -> int:
        # dataset size might change at runtime
        if self._num_samples is None:
            return len(self.data_source)
        return self._num_samples

    def __iter__(self):
        n = len(self.data_source)
        if self.replacement:
            for _ in range(self.num_samples // 32):
                yield from torch.randint(high=n, size=(32,), dtype=torch.int64, generator=self.generator).tolist()
            yield from torch.randint(high=n, size=(self.num_samples % 32,), dtype=torch.int64, generator=self.generator).tolist()
        else:
            yield from torch.randperm(n, generator=self.generator).tolist()

    def __len__(self):
        return self.num_samples
```

Reply:
IMO, it's better to keep it in `__iter__` to make RNG lazy initialized.

Reply:
In commit https://github.com/pytorch/pytorch/pull/53085/commits/60d010865f10897353aa7a2e8c1b5af073fc754c , I modified this part of the code.
But I realized that there was no need to construct a separate `genenrator` here, so I committed the changes again (https://github.com/pytorch/pytorch/pull/53085/commits/4645b7665861037676ceb372d201a9453ef0b4b7). 
I'm sorry that I didn't fill in the details in the commit massage here because I was in a hurry to commit.